libioc/Provisioning/puppet.py

```diff
@@ -0,0 +1,237 @@
+# Copyright (c) 2017-2019, Stefan Grönke, Igor Galić
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted providing that the following conditions
+# are met:
+# 1. Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""iocage provisioner for use with `puppet apply`."""
+import typing
+import os.path
+import urllib.parse
+
+import git
+
+import libioc.errors
+import libioc.events
+import libioc.Pkg
+import libioc.Provisioning
+
+
+class ControlRepoUnavailableError(libioc.errors.IocException):
+    """Raised when the puppet control-repo is not available."""
+
+    def __init__(
+        self,
+        url: str,
+        reason: str,
+        logger: typing.Optional['libioc.Logger.Logger']=None
+    ) -> None:
+        msg = f"Puppet control-repo '{url}' is not available: {reason}"
+        libioc.errors.IocException.__init__(
+            self,
+            message=msg,
+            logger=logger
+        )
+
+
+class ControlRepoDefinition(dict):
+    """Puppet control-repo definition."""
+
+    _url: str
+    _pkgs: typing.List[str]
+
+    def __init__(
+        self,
+        url: urllib.parse.DefragResult,
+        logger: 'libioc.Logger.Logger'
+    ) -> None:
+        self.logger = logger
+
+        if isinstance(url, urllib.parse.ParseResult) is False:
```

Comment:
so, this:
```sh-session
root@container-host1 /u/l/s/i/.libioc# ioc list -o name,provisioning.source webirc
+--------+-----------------------+
|  NAME  |  PROVISIONING.SOURCE  |
+========+=======================+
| webirc | /usr/local/etc/puppet |
+--------+-----------------------+
```

no longer works:

```python-traceback
root@container-host1 /u/l/s/i/.libioc# ioc -d spam provision webirc
/etc/rc.conf was read from /etc/rc.conf
Updated /etc/rc.conf data from /etc/rc.conf
Querying all running jails status
Executing: /usr/sbin/jls -v --libxo=json
    {"__version": "2", "jail-information": {"jail": [{"jid":15,"hostname":"webproxy","path":"/iocage/jails/webproxy/root","name":"default-webproxy","state":"ACTIVE","cpusetid":2, "ipv4_addrs": [], "ipv6_addrs": []}, {"jid":16,"hostname":"quassel","path":"/iocage/jails/quassel/root","name":"default-quassel","state":"ACTIVE","cpusetid":5, "ipv4_addrs": [], "ipv6_addrs": []}, {"jid":21,"hostname":"webirc","path":"/iocage/jails/webirc/root","name":"default-webirc","state":"ACTIVE","cpusetid":3, "ipv4_addrs": [], "ipv6_addrs": []}]}
    }
Injecting pre-loaded state to 'webirc'
[-] JailProvisioning@webirc: ...
  [-] JailProvisioningAssetDownload@webirc: FAILED [0.002s]
Traceback (most recent call last):
  File "/usr/local/bin/ioc", line 10, in <module>
    sys.dd:exit(cli())
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func                                                                                    
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/src/ioc/ioc_cli/provision.py", line 65, in cli
    **start_args
  File "/usr/local/src/ioc/ioc_cli/provision.py", line 102, in _provision
    print_function(_execute_provisioner(jail))
  File "/usr/local/src/ioc/ioc_cli/__init__.py", line 87, in print_events
    for event in generator:
  File "/usr/local/src/ioc/ioc_cli/provision.py", line 123, in _execute_provisioner
    for event in jail.provisioner.provision():
  File "/usr/local/lib/python3.6/site-packages/libioc/Provisioning/__init__.py", line 107, in provision                                                                      
    yield from self.__provisioning_module.provision(self)
  File "/usr/local/lib/python3.6/site-packages/libioc/Provisioning/puppet.py", line 182, in provision                                                                        
    raise e
  File "/usr/local/lib/python3.6/site-packages/libioc/Provisioning/puppet.py", line 177, in provision                                                                        
    logger=self.jail.logger
  File "/usr/local/lib/python3.6/site-packages/libioc/Provisioning/puppet.py", line 68, in __init__                                                                          
    raise TypeError("Source must be an URL")
TypeError: Source must be an URL
```
despite the code further down to support it.

Reply:
we should check for both AbsolutePath and URL

but we also have to catch whatever we throw here in the cli.

Reply:
> but we also have to catch whatever we throw here in the cli.

Why does the CLI need to check that?

Reply:
it needs to expect errors from the lib and print those, rather than a stack

Reply:
Then there should be an `IllegalSourceError` that the library provides. Errors inheriting from `IocException` should cause the CLI to `exit(1)` rather than printing a stack trace.

Reply:
this then should be the case for every error that happens at the edge of this library